# Worksheet 1

Welcome to the course 'Research Skills: Programming'! You're looking at Worksheet 1. It contains both informative text and exercises and if you work through each week's material, you will pass the assignments and hopefully also prepare you sufficiently for the exam that makes up the final grade.

Hopefully, you are reading this as a Jupyter Notebook in your browser. Most browsers work fine, though Internet Explorer sometimes seems to give problems. If you get into problems (for instance, you get a blank page instead of the notebook text) and you are running Internet Explorer, switch to another browser. If you are somehow reading this in Notepad, or found some other strange way to read this file... I want to say that I'm impressed first of all, but also make sure to install Anaconda Distribution over at: https://www.anaconda.com/products/distribution

Now, we'll need to talk a tiny bit about Python Notebooks. It's a special format for intermingling Python code and text; after installing the correct add-ons, Jupyter can automatically convert a .ipynb file into a nicely formatted PDF. But in this course, we'll mostly be using it as a kind of interactive notebook.

The notebooks consist of text interspersed with code. For instance, the greyish box below is some code that may look familiar from the lecture:

In [ ]:
## Code block 1: Introducing ourselves.

print("Hello, world!")

You can run this code by clicking on it so that it gets selected, and then press Ctrl+Enter. Alternatively, you can click on the button at the top of the page with the “play” symbol. If you do either of these things, you should see the text "Hello, world!" appear beneath the code. Did that work? (I'm assuming "yes") Great! Anything that a piece of code prints, is shown beneath it.

You even have the ability to change the code, and see what happens if you change it. For instance, in the code above, change the text “Hello, world!” to something else (but make sure you leave the double quotes in place, and do not use double quotes in your own text). Run the code, and see how it displays your own text. Be aware, however, that this actually changes your notebook; you cannot simply revert back to the original version of the notebook, unless you upload or copy a “clean” version. You can make copies and/or create “checkpoints” for the notebook in the “File” menu.

Note: Most students appreciate that notebooks store the changes that they make to them, but are feeling a bit uncertain about leaving a notebook without actually pressing a save button somewhere. Luckily, such a save button exists: you find it as the most left of the icons at the top of the page.

If for some reason you want to clear the output that you generated by running code, you can click on “Cell” in the menu at the top of the screen, then select “All Output” and “Clear”. Or, if you just want to clear the output of one piece of code, select it by clicking on it, then choose “Cell”, “Current Output”, “Clear”.

By the way, it is also possible for you to change the texts of the notebooks. When you double click on a text, or sometimes accidentally “scroll” into the text from code, suddenly the text is shown on a grey background and with different formatting. You can return the text to the regular visualization by “running it” (pressing Ctrl+Enter). I would love to lock the texts, but in the current version of the notebooks that is not possible. I am sure that we get that feature in the future, but for now we have to make do.

In general, in this course, each .ipynb Worksheet will have exercises scattered throughout the text, as well as a set of exercises at the end of each section. As you will also see, the first line in the code block gives some title information (that is the line starting with `##`). Exercises are indicated in bold with square braces like so **[Exercise X]**, and a code chunk will be provided for you to type your answer into. So always type your answers in the code chunks, either as a piece of code, or as comments (which will be covered in a bit).

Before we continue with the material, I want to give some information that will hopefully get you in the right mindset if you are unfamiliar with programming. The programming course is really like no other course you have seen in Communication and Information Sciences. The way you learn how to program is by failing a bunch of times until the code finally does what you want it to do. So don't get discouraged when things do not work! When the output is "wrong" or you get an error, it does not mean you are bad at programming, it just means that you are working towards the right answer. :) What is very typical is that you work for ages on one of the more challenging exercises, and then when you finally get the right anwer, you think "Ah jeez, all that work and it was just a few lines of code." That's totally normal and part of the process of learning how to program! As I said in the lecture: it is all a puzzle and puzzles are fun. The most important skill when you work on a challenging exercise is to not feel overwhelmed by it, but rather try to break the exercise down into its core components. When you know the steps that you need to undertake to finish an exercise, the hardest part is over. Also important to note: Google is your friend! Functions are explained in more detail on the Python website, and the website [Stack Overflow](https://stackoverflow.com) probably has the answer to all of your issues (well, not \*all\* of your issues, but you know what I mean). So if you add `site:stackoverflow.com` to your Google query, you should be able to find someone who was experiencing a similar issue, and the way to solve it. Figuring out the best way to Google your issues is an important programming (and life) skill.

### Calculations, Variables, Operators, and Comments

First things first, let's talk about some basics of Python. As you have seen in the lecture, its most straightforward use is as a calculator. Type any calculation into the code block below, hit Ctrl+Enter, and Jupyter will produce a result (try it!)

In [ ]:
## Code block 2: Python as a calculator

1 + 1

Note: one of the nice things with Jupyter notebooks is that just typing the name of a variable, or a calculation is enough to print the result (= evaluation) below the code block (also called "printing to the console"). When you are working with a different format (e.g., Python files), this will not work. In that case, you need to put what you want to print inside a print-statement, i.e. `print()`:


In [ ]:
## Code block 3: Python as a calculator with print statement

print(1 + 1)

As you see, code block 3 outputs the exact same thing as code block 2. This print statement is not necessary when you only want to print one evaluation in your code block, but oftentimes we want to print more than one thing, so it is good to familiarize yourself with the idea that you write a print statement around what you want to print.

You can also store things in variables, using the `=` operator. The name of the variable goes to the left of the equals-sign, its contents go to the right. Then if you want to see what is stored in the variable, type its name in the code block. For example:

In [ ]:
## Code block 3: Variables

my_var = "a"
print(my_var)

This has put the character `"a"` inside the variable `my_var`. You can store anything as a variable: integers, floating-point numbers, strings, `True` and `False`, as well as more complex data structures, like lists and functions, that we will get to later. Feel free to try to store other values in the `my_var` variable in the code block above. You can change an object's type at will; re-using it just overwrites it:

In [ ]:
## Code block 4: Overwriting a variable

my_var = 1
print(my_var)
my_var = True
print(my_var)

Also, if you like, you can try to change the values in `my_var` in the code block above and remove the `print()` statements, to see what happens if you don't use those.

If you want Python to store a string like `"a"` or maybe `"apple"`, you have to surround it in quotation marks (both single and double quotation marks work; some people have very strong opinions on what is best, I don't). Otherwise Python thinks you're referring to the name of an object. So this will give an error:

In [ ]:
## Code block 5: Strings require quotation marks

my_var = apple

If you see errors like this in the code block examples, try to make a mental note of it and what it means. NameError in this case means that we are trying to work with a variable (`apple`) that has not been created. In the exercises you may come across similar errors (which, again, is totally normal!) and it is very useful to have an idea of what they mean and how to solve them.

You can give basically any name to variables, but it is always good to choose a descriptive name that helps you and others understand what is stored (think of names like `age`, `car_brand`, `musician_name`, etc.). There are a few rules though:

- A variable name must start with a letter or the underscore character
- A variable name cannot start with a number
- A variable name can only contain alpha-numeric characters and underscores (`A-z`, `0-9`, and `_` )
- Variable names are case-sensitive (`age`, `Age` and `AGE` are three different variables)

**[Exercise 1]** Keeping the above information in mind, create your own object and print it.

In [2]:
## Exercise 1: Create your own variable and print it

my_first_object = "Hello!"

print(my_first_object)

Hello!


By the way, hopefully you have noticed that I have used `##` at the start of all blocks: That is a signal to Python that everything that follows on the same line is a comment, not actual code. This can also be nice for your own code sometimes, to explain to future-you what is actually happening. A single `#` is enough to signal that the line is a comment. The general convention is that you use just a single `#` to mean "this is a useful comment" and `##` to mean "I don't think this is useful, but I'll keep it here because you never know. In any case, pretend it doesn't even exist".

**[Exercise 2]** Take a look at the code blocks below. It contains variables with information. Try to predict the result of the expressions in the code blocks. What will you see when you run the code blocks? Add a comment to each of the two code blocks below, explaining each result.

In [3]:
## Exercise 2a: What is up with these variables?

1_car = True

print(1_car)

SyntaxError: invalid decimal literal (Temp/ipykernel_9060/2362271898.py, line 3)

In [5]:
## Exercise 2b: What is up with these variables?

_my_car = 'Toyota'

_My_Car = 'Mercedes'

print(_my_car)



Toyota


## Expressions

In the lecture, we have also discussed the ways that Python is more than just a simple calculator and how there is a certain "grammar" for expressions.

**[Exercise 3]** In the code block below, we have examples of expressions that do not "adhere to the grammar" resulting in code that is giving us errors. Can you fix these expressions so that they do not throw errors anymore and print the right results?

In [9]:
## Exercise 3: Fixing errors in expressions

expression1 = 3 + 6
print(expression1)
expression2 = '3' + ' birds'
print(expression2)
expression3 = 12 / 4
print(expression3)

9
3 birds
3.0


## Exercises

**[Exercise 4]** Let's take the following text:

``My favorite drink is: x``
``My least favorite drink is: y``

Can you make a script where we can freely change the values of `x` and `y`, and we print the text with the values for x and y as outcome. So when we use the value ``Klok`` for `x`, and ``Heineken`` for `y`, we get the output:

``My favorite drink is: Klok``
``My least favorite drink is: Heineken``

In [11]:
## Exercise 4: Modifying texts.

x = 'Coca Cola'
y = 'Pepsi'

favorite_drink = 'My favorite drink is: ' + x
least_favorite = 'My least favorite drink is: ' + y
print(favorite_drink)
print(least_favorite)

My favorite drink is: Coca Cola
My least favorite drink is: Pepsi


**[Exercise 5]** For this exercise, for some reason, you want to convert bananas to weight in kilograms. 1 banana weighs about 0.113 kilograms, so the formula is `x * 0.113` where `x` is the number of bananas, and we'd like to be able to freely change this value of `x` (you are free to assume that only integers will be stored in `x`). The final output (that should be printed) is the result of this calculation.

In [13]:
## Exercise 5: Banana for scale

bananas = 5

banana_calculation = bananas * 0.113
print(banana_calculation)

0.5650000000000001


## Comparison operators, Boolean values, and flow control statements

In many cases, we do not want to just perform basic calculations and modifications with Python. We want to use this calculated/modified information and process it in some way. For instance, by comparing values to each other. That's where comparison operators in combination with Boolean values come into play. As said in the lecture, there are two Boolean values: `True` or `False`, and comparisons that we are making will boil down to either one of them. Let's check it out.


In [2]:
## Code block 6: Boolean comparisons

firstcomparison = 4 > 3
print(firstcomparison)

secondcomparison = 4 < 3
print(secondcomparison)

True
False


In the code block above, we first ask _"Is the value `4` more than the value `3`?"_, for which the answer is _"yes"_, so it takes on the Boolean value `True`. We than ask _"Is the value `4` less than the value `3`?"_, which is not the case, of course: `False`.

We can extend the amount of processing we do even further, by using _Flow control statements_. Then, we don't just want our yes/no (`True`/`False`) answer, but we also want to do something with that `True`/`False` answer. This is done with `if`, `elif`, and `else`. In this worksheet, we will focus on the `if` and `else` statements. Let's compare the following code block to code block 6.

In [4]:
## Code block 7: Flow control statements

if 4 > 3:
    print('4 is indeed more than 3')
else:
    print('No, 4 is less than 3')

4 is indeed more than 3


As you may be able to tell, code block 7 is basically an extension of the `firstcomparison` variable of code block 6. But rather than printing (`True`/`False`), we print a more detailed string. What this code block basically does is it starts again with the question _"Is the value `4` more than the value `3`?"_, but then extends it with an _"If so, print `4 is indeed more than 3` to the console (the `if`-statement), but if that is not the case, print `No, 4 is less than 3` to the console (the `else` statement). Pay special attention to the spacing that is applied here. When you are working with flow control statements, you want everything that belongs to the `if`, `elif`, or `else` to be bundled together by a tab (or 4 spaces) at the start of the line. Fierce discussions have been had between programmers about whether using a tab or 4 spaces is better (I wish I was kidding...), but both are fine in my book. As long as you pay very close attention to the spacing.

In [5]:
## Code block 8: Pay attention the spacing

if 3 < 4:
    print('3 is indeed less than 4')
    print('That speaks for itself')
    print('Why do you even ask such a question?')
    print('Are you bored or something?')
else:
    print('No, 3 is more than 4')
    print('Math has been revolutionized')
    print('Have you ever heard of the actor Terrence Howard?')
    print('He has invented a type of math where this might be true.')
    print('He was right all along.')

3 is indeed less than 4
That speaks for itself
Why do you even ask such a question?
Are you bored or something?


As you see here, there is kind of a "text island" below the if-statement and else-statement. There are multiple print statements that are all one tab up from the start of the line. This bundles all the print lines together below the if-statement, and else-statement respectively. Feel free to change the `3 < 4` with another comparison expression to also see the lines below `else` in action.

**[Exercise 6]** Why are the comparison expressions below not both `True`? Write your answer as a comment at the bottom of the code block.

In [6]:
## Exercise 6: A tale of two comparison expressions

comparisonexp1 = 5 >= 5
print(comparisonexp1)

comparisonexp2 = 5 > 5
print(comparisonexp2)

True
False


**[Exercise 7]** We're going to re-use the formula from Exercise 5, where for some reason, you want to convert bananas to weight in kilograms. As a reminder: 1 banana weighs about 0.113 kilograms, so the formula is `x * 0.113` where `x` is the number of bananas, and we'd like to be able to freely change this value of `x`. We want to store the total weight in kilograms in the variable `bananas`.

So, you have a variable `x`, in which we store the integer `5` for example. This means, we are calculating `5 * 0.113 = 0.565`. The value `0.565` is then stored in the variable `bananas`.

If this value of `bananas` is more than `1`, print the string `"Wow, those are too many bananas to carry"`. If it is not more than `1`, print the string `"I should be able to carry this home"`.

**[Bonus exercise 1]** Bonus exercises are totally optional! Only do this exercise if you feel like doing more programming. Change the comparison to something other than "more than 1" vs. "not more than 1". Add an `elif` statement into the mix.

**[Bonus exercise 2]** Same here, totally optional. Change the strings that are printed. They should take the form of `"[x] bananas weigh [bananas] total. Too heavy."` or `"[x] bananas weigh [bananas] total. I can carry this."` Where `x` and `bananas` correspond with the values of the variables `x` and `bananas`. So if you change the value of `x`, the output of `x` and `bananas` should change as well. Hint: look up how you can change data types.

In [19]:
## Exercise 7: Banana for scale (pt. 2)

x = 100

bananas = x * 0.113

if bananas > 1:
    print("Wow, those are too many bananas to carry")
elif bananas <= 1:
    print("I should be able to carry this home")

##Or you could use "else:"

if bananas > 1:
    print("Wow, those are too many bananas to carry")
else:
    print("I should be able to carry this home")

Wow, those are too many bananas to carry


That is all! If you run into any issues, do not forget to ask about this on the Discussion board, or in the Practical session!

## Skills & Further Resources

This worksheet has introduced the basics of Python programming.

After working through it, you should be able to:

1. Print output to the console
2. Use Python as a calculator
3. Store information inside a variable
4. Understand the rules for the names you can give variables
5. Understand the difference between strings, integers, and floating-point numbers
6. Modify variables

If you want a bit more information on some of these topics, besides the chapters in Automate the Boring Stuff, I recommend:

- The Introduction to Python course on https://datacamp.com ; more specifically, Chapter 1: Python Basics

## Overview of New Information

| Python Code                |                                 does what                                  |
|----------------------------|:--------------------------------------------------------------------------:|
| `x + y`                    |     Basic calculations (other arithmetic operators are also possible)      |
| `print()`                  |              Print the output of an expression to the console              |
| `x = y`                    |                          Assign `y` to object `x`                          |
| `1` vs `1.0003` vs `"one"` | Formatting of integers, floating-point numbers, and strings (respectively) |
| `x > y`                    |     Compare `x` and `y` (other comparison operators are also possible)     |
| `True` and `False`         |        Boolean values that are generally the result of a comparison        |
| `if`, `elif`, and `else`   | Flow control statements to react differently to the result of a comparison |